### PR TITLE
Sara/mkt 279 update ignore operations nosemgrep

### PR DESCRIPTION
**Pull request by s-santillan**
### Security

- [x] Change has no security implications (otherwise, ping the security team)

### CODE REVIEW

**Review by ievans — COMMENTED** [review]
We will need a redirect, as I believe the CLI and app reference the `ignoring findings` page by URL.
Overall the new page feels quite long.

As previously mentioned, the content is great!

**Comment by s-santillan** [comment]
> We will need a redirect, as I believe the CLI and app reference the `ignoring findings` page by URL. Overall the new page feels quite long.
> 
> As previously mentioned, the content is great!

@ievans Thank you! I am going through the notes now. With regards to the redirect, it has been set up in docusaurus.config.js in a previous commit. To correctly rebuild and test, `yarn build` then `yarn serve` fixes it for me. Noted on the length. Will see if it can be condensed further or have some kind of rich media (image, graph, figure, etc.) in the future to break up the text.

File: docs/ignoring-files-folders-code.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+
+All Semgrep environments (CLI, CI, and App) adhere to user-defined or Semgrep-defined ignore patterns.
+
+
+## Reference summary
+
+| Method  | Usage    | Examples |
+|:------- |:-------- |:-------- |
+| To ignore blocks of code: `nosemgrep` | Create a comment, followed by a space (` `), followed by `nosemgrep` at the first line or preceding line of the pattern match. | ` // nosemgrep` &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; <br /> `# nosemgrep` |
+| To ignore files and folders: `.semgrepignore` | Create a `.semgrepignore` file and add patterns for files and folders there. Patterns follow `.gitignore` syntax with some caveats. See [Defining files and folders in `.semgrepignore`](#defining-files-and-folders-in-semgrepignore). | `.gitignore` <br /><br /> `.git` |
```
[the remark this answers is not in the document]
Comment (s-santillan) [reply docs/ignoring-files-folders-code.md]: I see the confusion about using .git, .gitignore as examples of the concept. I instead linked them to the actual semgrepignore file as a sample. Have also added the ` //nosemgrep: rule-id` example. Resolving...

File: docs/ignoring-files-folders-code.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+The following is **Semgrep's default `.semgrepignore` file**:
+
+```
+DEFAULT_SEMGREPIGNORE_TEXT
+```
+
+:::caution
+
+Note that the default `.semgrepignore` file is opinionated and causes Semgrep to skip these folders:
```
[the remark this answers is not in the document]
Comment (s-santillan) [reply docs/ignoring-files-folders-code.md:43]: When you say "that", do you mean patterns of entries like "tests", for example "foo_tests"? I used specific entries:
```
# Common test paths
test/
tests/
```

There is a pattern,  `*_test.go` — do you think it's worth it to add? The goal of the caution message is to make sure people read the .semgrepignore file or refer to it when they need to debug.

File: docs/ignoring-files-folders-code.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+title: Ignoring files, folders, or code
+description: "This documents various methods to skip or ignore files, folders, or code that are not relevant to a Semgrep scan."
+---
+
+import MoreHelp from "/src/components/MoreHelp"
+
+
+# Ignoring files, folders, or code
+
+**Ignoring files, folders, or code** means Semgrep will not scan certain defined items within your repository or working directory. Thus no findings are generated.
```
Comment (pabloest) [inline docs/ignoring-files-folders-code.md:13]:
```suggestion
**Ignoring files, folders, or code** means Semgrep will not scan certain defined items within your repository or working directory. Thus no findings are generated from the ignored items.
```

File: docs/ignoring-files-folders-code.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+The following is **Semgrep's default `.semgrepignore` file**:
+
+```
+DEFAULT_SEMGREPIGNORE_TEXT
+```
+
+:::caution
+
+Note that the default `.semgrepignore` file is opinionated and causes Semgrep to skip these folders:
```
Comment (pabloest) [inline docs/ignoring-files-folders-code.md:43]:
```suggestion
The default `.semgrepignore` file is opinionated and causes Semgrep to skip these folders:
```

File: docs/ignoring-files-folders-code.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+|:--------|:---------|:---------|
+| To ignore blocks of code: `nosemgrep` | Create a comment, followed by a space (` `), followed by `nosemgrep` at the first line or preceding line of the pattern match. | ` // nosemgrep` &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; `// nosemgrep: rule-id` <br /> `# nosemgrep` |
+| To ignore files and folders: `.semgrepignore` | Create a `.semgrepignore` file and add patterns for files and folders there. Patterns follow `.gitignore` syntax with some caveats. See [Defining files and folders in `.semgrepignore`](#defining-files-and-folders-in-semgrepignore). | [`.semgrepignore` sample file](https://raw.githubusercontent.com/returntocorp/semgrep/develop/semgrep/semgrep/templates/.semgrepignore) |
+
+
+## Understanding Semgrep defaults
+
+Without user customization, Semgrep refers to the following to define ignored files and folders:
+
+* Semgrep's default `.semgrepignore` file
```
Comment (pabloest) [inline docs/ignoring-files-folders-code.md:32]: One thing that is not clear to me about the .semgrepignore file is where it comes from and where it should be placed. As written here, I could understand that Semgrep (the program) has a .semgrepignore file and as a user I get it when I install Semgrep. But I think that’s not the case? I think this default file refers to the template we have in the Semgrep, repo, right?

If I understand correctly, then the user does need to take some action (perhaps not customization, strictly speaking) to make use of this default file.

File: docs/ignoring-files-folders-code.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+This document defines **files, folders and code** as those that are **relevant to a Semgrep scan**. For example, `.jpg` files are not a part of Semgrep's scope and therefore are not part of the scope of this document.
+
+
+## Customizing ignore behavior
+
+Semgrep provides several methods to customize ignore behavior. Refer to the following table to see which method suits your goal:
+
+| Goal | Method |
+|:---- |:------ |
+| To scan all files each time you run Semgrep (no files will be ignored) | Create an empty `.semgrepignore` file. An empty `.semgrepignore` will result in patterns within `.gitignore` **to be scanned**. |
```
Comment (pabloest) [inline docs/ignoring-files-folders-code.md:70]: Where should this .semgrepignore file be placed?
(maybe it’s better to state its required location elsewhere, and not in this table)

File: docs/ignoring-files-folders-code.md [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+| To ignore files and folders in `.git` and `.gitignore`. | Add `.git` and `.gitignore` to your `.semgrepignore` file. |
+| To ignore custom files and folders each time you run a scan. | Add these files to your `.semgrepignore` file or define them through Semgrep App.|
+| To ignore specific code blocks each time you run a scan. | Create a comment with the word `nosemgrep`. |
+| To ignore files or folders for a particular scan. | Run Semgrep with the flag `--exclude` followed by the pattern or file to be excluded. See CLI reference.
```
Comment (pabloest) [inline docs/ignoring-files-folders-code.md:74]:
```suggestion
| To ignore files or folders for a particular scan. | Run Semgrep with the flag `--exclude` followed by the pattern or file to be excluded. See [CLI reference](../cli-usage/).
```

File: docs/ignoring-files-folders-code.md [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+1. Sign into [Semgrep App](https://semgrep.dev/login?return_path=/manage/projects).
+2. From the Dashboard Sidebar, select **Projects > [Project name]**.
+3. Enter files and folders to ignore in the **Path Ignores** box.
+
+Including files and folders through this method is **additive**. When Semgrep App makes a scan, it will look for a `.semgrepignore` within the repository. If no `.semgrepignore` file is found, it will temporarily create one and add items from Semgrep App's Path Ignores. Adding items to the **Path Ignores** box will not override default Semgrep ignore patterns.
+
+
+## Ignoring code through nosemgrep
+
+To ignore blocks of code, define an **inline comment**, followed by a **space** ( ), followed by the word `nosemgrep` at either the **first line** or **the line preceding** the potential match. Semgrep will ignore all rule pattern matches. This functionality works across all supported languages.
```
Comment (pabloest) [inline docs/ignoring-files-folders-code.md:110]: On line 24 the space is written as:

…followed by a space (` `)…

File: docs/ignoring-files-folders-code.md [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+
+bad_func2(   // nosemgrep
+    arg
+);
+
+```
+
+:::info
+
+The space ( ) before `nosemgrep` is required for Semgrep to detect this annotation.
```
Comment (pabloest) [inline docs/ignoring-files-folders-code.md:137]: Same comment as the comment for line 110

**Review by pabloest — APPROVED** [review]

File: docs/ignoring-files-folders-code.md [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+## Customizing ignore behavior
+
+Semgrep provides several methods to customize ignore behavior. Refer to the following table to see which method suits your goal:
+
+| Goal | Method |
+|:---- |:------ |
+| To scan all files each time you run Semgrep (no files will be ignored) | Create an empty `.semgrepignore` file. An empty `.semgrepignore` will result in patterns within `.gitignore` **to be scanned**. |
+| To ignore files and folders in `.git` and `.gitignore`. | Add `.git` and `.gitignore` to your `.semgrepignore` file. |
+| To ignore custom files and folders each time you run a scan. | Add these files to your `.semgrepignore` file or define them through Semgrep App.|
```
[the remark this answers is not in the document]
Comment (s-santillan) [reply docs/ignoring-files-folders-code.md:72]: Committed. Resolving.

File: docs/ignoring-files-folders-code.md (hunk 6 above)
Comment (s-santillan) [reply docs/ignoring-files-folders-code.md]: I mention it twice in my update (both tables) so people don't miss it :) Thanks for the catch!

File: docs/ignoring-files-folders-code.md (hunk 5 above)
Comment (s-santillan) [reply docs/ignoring-files-folders-code.md:32]: I've replaced
> Semgrep's default `.semgrepignore` file:

with
>In the absence of a user-generated `.semgrepignore`, Semgrep will refer to its repository's `.semgrepignore` default template:

To make it explicit "where" this default is coming from. Please let me know if this makes sense?

Comment (pabloest) [reply docs/ignoring-files-folders-code.md:32]: Yes, thank you!

File: docs/ignoring-files-folders-code.md [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+
+
+## Customizing ignore behavior
+
+Semgrep provides several methods to customize ignore behavior. Refer to the following table to see which method suits your goal:
+
+| Goal | Method |
+|:---- |:------ |
+| To scan all files each time you run Semgrep (no files will be ignored) | Create an empty `.semgrepignore` file. An empty `.semgrepignore` will result in patterns within `.gitignore` **to be scanned**. |
+| To ignore files and folders in `.git` and `.gitignore`. | Add `.git` and `.gitignore` to your `.semgrepignore` file. |
```
[the remark this answers is not in the document]
Comment (s-santillan) [reply docs/ignoring-files-folders-code.md]: "doesn't semgrep implicitly ignore `.git` unless you pass the `no-git-ignore`" 

I tested this several times to be sure. .git and .gitignore are handled differently. 

.gitignore is scanned if a user-created .semgrepignore is empty. `git` is still skipped. I tested this by adding the slowest file to scan in my .gitignore, creating an .emptysemgrepignore, and counting the files scanned at the end. Counting the files verified that .git was skipped, but .gitignore was included. 

I have noted this down in the latest updates. Resolving.

File: docs/ignoring-files-folders-code.md (hunk 9 above)
Comment (s-santillan) [reply docs/ignoring-files-folders-code.md]: Updated in latest docs. Resolving both tickets.

**Pull request merged**